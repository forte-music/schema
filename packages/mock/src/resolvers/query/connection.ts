import { Connection } from '../../models';

export interface ConnectionArgs {
  after?: string;
  first?: number;
}

export const handleConnection = <InputType, NodeType>(
  keys: InputType[],
  getNode: (key: InputType) => NodeType,
  { first = 25, after }: ConnectionArgs
): Connection<NodeType> => {
  const lowerBound = after ? parseInt(after, 10) + 1 : 0;
  const upperBound = first === -1 ? keys.length : lowerBound + first;

  const acceptedKeys = keys.slice(lowerBound, upperBound);

  return {
    count: keys.length,
    pageInfo: {
      hasNextPage: upperBound < keys.length,
    },
    edges: acceptedKeys.map((key, index) => ({
      cursor: (index + lowerBound).toString(),
      node: getNode(key),
    })),
  };
};
