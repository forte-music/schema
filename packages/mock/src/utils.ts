type Identifiable = {
  id: string;
};

export const mustGet = <K, V>(map: Map<K, V>, key: K): V => {
  const value = map.get(key);

  if (value === undefined) {
    throw new TypeError(`Schema invalid. Value not found in map. Key: ${key}`);
  }

  return value;
};

export const mustGetKeys = <T>(map: Map<string, T>, keys: string[]): T[] =>
  keys.map(key => mustGet(map, key));

// The getMap is needed because due to circular imports, the map may be
// undefined at import time. For example if song imports album's map and
// album import's songs map album's song map will be undefined because song
// hasn't exported anything yet.
export const arrayPropertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  keys: string[]
) => ({
  get: () => mustGetKeys(getMap(), keys),
});

export const propertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
});

export const propertyDescriptorWithSet = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
  set: (val: T): void => {
    getMap().set(key, val);
  },
});

export const makeMap = <T extends Identifiable>(list: T[]): Map<string, T> =>
  list.reduce((map, identifiable: T) => {
    addToMap(map, identifiable);
    return map;
  }, new Map());

export const addToMap = <T extends Identifiable>(
  map: Map<string, T>,
  item: T
) => {
  map.set(item.id, item);
};

export const now = () => Math.floor(Date.now() / 1000);

export const uuidForNum = (num: number): string =>
  num
    .toString(16)
    .toLowerCase()
    .padStart(32, '0');
