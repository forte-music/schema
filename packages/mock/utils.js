// @flow
type Identifiable = {
  id: string,
};

export const mustGet = <K: string, V>(map: Map<K, V>, key: K): V => {
  const value = map.get(key);

  if (value === undefined) {
    throw new TypeError(`Schema invalid. Value not found in map. Key: ${key}`);
  }

  return value;
};

export const mustGetKeys = <T>(map: Map<string, T>, keys: string[]): T[] =>
  keys.map(key => mustGet(map, key));

// The getMap is needed because due to circular imports, the map may be
// undefined at import time. For example if song imports album's map and
// album import's songs map album's song map will be undefined because song
// hasn't exported anything yet.
export const arrayPropertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  keys: string[]
) => ({
  get: () => mustGetKeys(getMap(), keys),
});

export const propertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
});

export const propertyDescriptorWithSet = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
  set: (val: T): void => {
    getMap().set(key, val);
  },
});

export const makeMap = <T: Identifiable>(list: T[]): Map<string, T> =>
  list.reduce((map, identifiable: T) => {
    map.set(identifiable.id, identifiable);
    return map;
  }, new Map());
