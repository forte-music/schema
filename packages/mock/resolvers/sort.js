// @flow
import type { Connection, UserStats } from '../models';
import type { ConnectionArgs } from './connection';
import { handleConnection } from './connection';

type SortBy =
  | 'RECENTLY_ADDED'
  | 'LEXICOGRAPHICALLY'
  | 'RECENTLY_PLAYED'
  | 'RELEVANCE';

type SortParams = {
  sortBy?: SortBy,
  reverse?: boolean,
  filter?: string,
};

type Sortable = {
  name: string,
  stats: UserStats,
  timeAdded: number,
};

type LinkedSortInfo<T> = {
  sortable: Sortable,
  original: T,
};

const compare = (a: any, b: any): number => {
  if (a > b) {
    return 1;
  }

  if (a === b) {
    return 0;
  }

  if (b > a) {
    return -1;
  }

  throw new TypeError(`Invariant a: ${a}, b: ${b}`);
};

// Returns how closely the record matches the query string.
const searchDistance = (record: string, query: string): number => {
  const normalizedQuery = query.toLowerCase().replace(/ /g, '');
  const normalizedRecord = record.toLowerCase();

  let strength = 0;
  let currentStrength = 0;
  let lastMatchIndex = 0;

  const khars = normalizedQuery.split('');
  for (const khar of khars) {
    const indexOfQueryCharacter = normalizedRecord.indexOf(
      khar,
      lastMatchIndex
    );
    if (indexOfQueryCharacter === -1) {
      // A character from the query isn't in the record. Sort to bottom.
      return -1;
    }

    if (lastMatchIndex === indexOfQueryCharacter - 1) {
      // continuation of last match
      currentStrength++;
    } else {
      // broken streak, reset streak strength
      currentStrength = 0;
    }

    // move last match index
    lastMatchIndex = indexOfQueryCharacter;

    // update strength
    strength += currentStrength;
  }

  return strength;
};

const reverseCompare = (a: any, b: any): number => reverse(a, b, compare);

const reverse = <T, R>(a: T, b: T, inner: (a: T, b: T) => R): R => inner(b, a);

const identity = <T, R>(a: T, b: T, inner: (a: T, b: T) => R): R => inner(a, b);

type Comparator<T> = (T, T) => number;

const sortWithReverse = <T>(
  reversed: boolean,
  inner: Comparator<T>
): Comparator<T> => {
  return (a: T, b: T): number => {
    const transform = reversed ? reverse : identity;
    return transform(a, b, inner);
  };
};

export type SortConnectionArgs = ConnectionArgs & {
  sort?: SortParams,
};

export const listItemsResolver = <T>(
  values: T[],
  { sort, ...connectionArgs }: SortConnectionArgs,
  process: (item: T) => Sortable
): Connection<T> => {
  const localHandleConnection = (items: T[]) =>
    handleConnection(items, value => value, connectionArgs);

  if (!sort) {
    // No sort was specified, use default sort.

    return localHandleConnection(values);
  }

  const { sortBy = 'LEXICOGRAPHICALLY', reverse = false, filter = '' } = sort;

  let sorter: Comparator<Sortable> = (a: Sortable, b: Sortable): number => 0;
  let filterFunc = (item: Sortable) => searchDistance(item.name, filter) !== -1;

  switch (sortBy) {
    case 'RECENTLY_ADDED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.timeAdded, b.timeAdded);
      break;

    case 'LEXICOGRAPHICALLY':
      sorter = (a: Sortable, b: Sortable) => compare(a.name, b.name);
      break;

    case 'RECENTLY_PLAYED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.stats.lastPlayed || 0, b.stats.lastPlayed || 0);
      break;

    case 'RELEVANCE':
      if (!filter) {
        throw new TypeError('Used relevance sort without filter.');
      }

      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(
          searchDistance(a.name, filter),
          searchDistance(b.name, filter)
        );

      break;

    default:
      throw new TypeError(`Unknown Sort ${sortBy}`);
  }

  const mappedValues: LinkedSortInfo<T>[] = values.map(item => ({
    original: item,
    sortable: process(item),
  }));

  const filteredValues = mappedValues.filter(({ original: T, sortable }) =>
    filterFunc(sortable)
  );

  const sortedValues = filteredValues.slice();
  sortedValues.sort(
    sortWithReverse(
      reverse,
      (
        { sortable: a }: LinkedSortInfo<T>,
        { sortable: b }: LinkedSortInfo<T>
      ) => sorter(a, b)
    )
  );

  return localHandleConnection(sortedValues.map(({ original }) => original));
};

export const itemsResolver = <T: Sortable>(map: Map<string, T>) => (
  _: void,
  args: SortConnectionArgs
): Connection<T> => {
  const values = Array.from(map.values());
  return listItemsResolver(values, args, item => item);
};
