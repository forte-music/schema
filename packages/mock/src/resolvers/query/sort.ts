import { Connection } from '../../models';
import { ConnectionArgs, handleConnection } from './connection';

type SortBy = 'RECENTLY_ADDED' | 'LEXICOGRAPHICALLY' | 'RECENTLY_PLAYED';

interface SortParams {
  sortBy?: SortBy;
  reverse?: boolean;
  filter?: string;
}

interface Sortable {
  name: string;
  lastPlayed?: number;
  timeAdded: number;
}

interface LinkedSortInfo<T> {
  sortable: Sortable;
  original: T;
}

interface Comparator<T> {
  (a: T, b: T): number;
}

export const compare = <T>(a: T, b: T): number => {
  if (a > b) {
    return 1;
  }

  if (a === b) {
    return 0;
  }

  if (b > a) {
    return -1;
  }

  throw new TypeError(`Invariant a: ${a}, b: ${b}`);
};

export const reverseCompare = <T>(a: T, b: T): number => reverse(a, b, compare);

const reverse = <T>(a: T, b: T, inner: Comparator<T>): number => inner(b, a);

const identity = <T>(a: T, b: T, inner: Comparator<T>): number => inner(a, b);

// Returns how closely the record matches the query string.
const searchDistance = (record: string, query: string): number => {
  const normalizedQuery = query.toLowerCase().replace(/ /g, '');
  const normalizedRecord = record.toLowerCase();

  let strength = 0;
  let currentStrength = 0;
  let lastMatchIndex = 0;

  const khars = normalizedQuery.split('');
  for (const khar of khars) {
    const indexOfQueryCharacter = normalizedRecord.indexOf(
      khar,
      lastMatchIndex
    );
    if (indexOfQueryCharacter === -1) {
      // A character from the query isn't in the record. Sort to bottom.
      return -1;
    }

    if (lastMatchIndex === indexOfQueryCharacter - 1) {
      // continuation of last match
      currentStrength++;
    } else {
      // broken streak, reset streak strength
      currentStrength = 0;
    }

    // move last match index
    lastMatchIndex = indexOfQueryCharacter;

    // update strength
    strength += currentStrength;
  }

  return strength;
};

const sortWithReverse = <T>(
  reversed: boolean,
  inner: Comparator<T>
): Comparator<T> => {
  return (a: T, b: T): number => {
    const transform = reversed ? reverse : identity;
    return transform(a, b, inner);
  };
};

export type SortConnectionArgs = ConnectionArgs & {
  sort?: SortParams;
};

export const listItemsResolver = <T>(
  values: T[],
  { sort, ...connectionArgs }: SortConnectionArgs,
  process: (item: T) => Sortable
): Connection<T> => {
  const localHandleConnection = (items: T[]) =>
    handleConnection(items, value => value, connectionArgs);

  if (!sort) {
    // No sort was specified, use default sort.

    return localHandleConnection(values);
  }

  const { sortBy = 'LEXICOGRAPHICALLY', reverse = false, filter = '' } = sort;

  let sorter: Comparator<Sortable> = () => 0;
  let filterFunc = (item: Sortable) => searchDistance(item.name, filter) !== -1;

  switch (sortBy) {
    case 'RECENTLY_ADDED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.timeAdded, b.timeAdded);
      break;

    case 'LEXICOGRAPHICALLY':
      sorter = (a: Sortable, b: Sortable) => compare(a.name, b.name);
      break;

    case 'RECENTLY_PLAYED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.lastPlayed || 0, b.lastPlayed || 0);
      break;

    default:
      throw new TypeError(`Unknown Sort ${sortBy}`);
  }

  const mappedValues: LinkedSortInfo<T>[] = values.map(item => ({
    original: item,
    sortable: process(item),
  }));

  const filteredValues = mappedValues.filter(({ sortable }) =>
    filterFunc(sortable)
  );

  const sortedValues = filteredValues.slice();
  sortedValues.sort(
    sortWithReverse(
      reverse,
      (
        { sortable: a }: LinkedSortInfo<T>,
        { sortable: b }: LinkedSortInfo<T>
      ) => sorter(a, b)
    )
  );

  return localHandleConnection(sortedValues.map(({ original }) => original));
};

export const itemsResolver = <T extends Sortable>(map: Map<string, T>) => (
  _: void,
  args: SortConnectionArgs
): Connection<T> => {
  const values = Array.from(map.values());
  return listItemsResolver(values, args, item => item);
};
