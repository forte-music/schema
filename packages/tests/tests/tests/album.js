import gql from 'graphql-tag';
import { print } from 'graphql/language/printer';

import { testConnection } from '../connection';
import { testSort } from '../sort';
import AlbumFields from './fragments/AlbumFields.graphql';
import ArtistFields from './fragments/ArtistFields.graphql';
import SongFields from './fragments/SongFields.graphql';
import UserStatsFields from './fragments/UserStatsFields.graphql';

import client from '../client';
import { uuidForNum } from '../utils';

const variables = { albumId: uuidForNum('4') };
it('should get an album by id', async () => {
  const query = gql`
    query($albumId: ID!) {
      album(id: $albumId) {
        ...AlbumFields

        artist {
          ...ArtistFields
        }

        songs {
          ...SongFields
        }
      }
    }

    ${AlbumFields}
    ${ArtistFields}
    ${SongFields}
  `;

  const { album } = await client.request(print(query), variables);

  expect(album).toMatchSnapshot();
});

it('should calculate the duration for an album', async () => {
  const { album } = await client.request(
    `
      query($albumId: ID!) {
        album(id: $albumId) {
          id
          duration
        }
      }
    `,
    variables
  );

  expect(album).toMatchObject({
    id: variables.albumId,
    duration: 74 * 60 + 22,
  });
});

testConnection('albums', async ({ first, after }) => {
  const query = gql`
    query($first: Int, $after: String) {
      albums(first: $first, after: $after) {
        count
        pageInfo {
          hasNextPage
        }

        edges {
          cursor

          node {
            id
          }
        }
      }
    }
  `;

  const { albums } = await client.request(print(query), {
    first,
    after,
  });

  return albums;
});

testSort('albums', async ({ sortBy, reverse }) => {
  const query = gql`
    query($sortBy: SortBy!, $reverse: Boolean!) {
      albums(first: -1, sort: { sortBy: $sortBy, reverse: $reverse }) {
        edges {
          node {
            id
            name
            timeAdded
            stats {
              ...UserStatsFields
            }
          }
        }
      }
    }

    ${UserStatsFields}
  `;

  const { albums: { edges } } = await client.request(print(query), {
    sortBy,
    reverse,
  });

  return edges.map(({ node }) => node);
});
