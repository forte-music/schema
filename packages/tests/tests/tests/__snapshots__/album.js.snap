// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get an album by id 1`] = `
Object {
  "artist": Object {
    "collageArtwork": Array [
      "https://upload.wikimedia.org/wikipedia/en/3/31/G-Eazy_-_The_Beautiful_%26_Damned.png",
      "https://upload.wikimedia.org/wikipedia/en/7/7a/G-Eazy_These_Things_Happen.jpg",
      "https://upload.wikimedia.org/wikipedia/en/d/d7/G-Eazy_When_Its_Dark_Out.jpg",
      "https://is5-ssl.mzstatic.com/image/thumb/Music71/v4/e6/c1/f8/e6c1f863-2269-2920-b5f0-74482fbbddd9/886446252223.jpg/1200x1200bb.jpg",
    ],
    "id": "00000000000000000000000000000004",
    "name": "G-Eazy",
  },
  "duration": 4462,
  "id": "00000000000000000000000000000004",
  "name": "The Beautiful & Damned",
  "releaseYear": 2017,
  "songs": Array [
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000004",
      "name": "The Beautiful & Damned",
      "trackNumber": 1,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000005",
      "name": "Pray for Me",
      "trackNumber": 2,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000006",
      "name": "Him & I",
      "trackNumber": 3,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000007",
      "name": "But a Dream",
      "trackNumber": 4,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000008",
      "name": "Sober",
      "trackNumber": 5,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000009",
      "name": "Legend",
      "trackNumber": 6,
    },
    Object {
      "diskNumber": 1,
      "id": "0000000000000000000000000000000a",
      "name": "No Limit",
      "trackNumber": 7,
    },
    Object {
      "diskNumber": 1,
      "id": "0000000000000000000000000000000b",
      "name": "The Plan",
      "trackNumber": 8,
    },
    Object {
      "diskNumber": 1,
      "id": "0000000000000000000000000000000c",
      "name": "That's a Lot",
      "trackNumber": 9,
    },
    Object {
      "diskNumber": 1,
      "id": "0000000000000000000000000000000d",
      "name": "Pick Me Up",
      "trackNumber": 10,
    },
    Object {
      "diskNumber": 1,
      "id": "0000000000000000000000000000000e",
      "name": "Gotdamn",
      "trackNumber": 11,
    },
    Object {
      "diskNumber": 1,
      "id": "0000000000000000000000000000000f",
      "name": "Leviathan",
      "trackNumber": 12,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000010",
      "name": "Crash & Burn",
      "trackNumber": 13,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000011",
      "name": "Summer in December",
      "trackNumber": 14,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000012",
      "name": "Charles Brown",
      "trackNumber": 15,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000013",
      "name": "No Less",
      "trackNumber": 16,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000014",
      "name": "Mama Always Told Me",
      "trackNumber": 17,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000015",
      "name": "Fly Away",
      "trackNumber": 18,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000016",
      "name": "Love Is Gone",
      "trackNumber": 19,
    },
    Object {
      "diskNumber": 1,
      "id": "00000000000000000000000000000017",
      "name": "Eazy",
      "trackNumber": 20,
    },
  ],
}
`;
